halo2/src/circuit_builder.rs

```diff
@@ -166,6 +170,38 @@ pub(crate) fn analyzed_to_circuit<T: FieldElement>(
             _ => unimplemented!(),
         }
     }
+
+    assert!(
+        analyzed.public_declarations.len() <= num_rows,
+        "More publics than rows!"
+    );
+
+    // Enforce publics by copy-constraining to cells in the instance column.
+    let mut copys = vec![];
+    let mut publics = vec![];
+
+    for public_declaration in analyzed.public_declarations.values() {
+        let witness_name = public_declaration.referenced_poly_name();
+        let witness_col = cd.col(&witness_name);
+        let witness_offset = public_declaration.index as usize;
+        let public_col = cd.public_column;
+        let public_offset = copys.len();
+
+        // Add copy constraint
+        copys.push(CopyC {
+            columns: (public_col, witness_col),
+            // We could also create one copy constraint per column pair wih several offsets.
+            // I don't thin there is a difference though...
+            offsets: vec![(public_offset, witness_offset)],
+        });
+
+        // Evaluate the given cell and add it to the publics.
+        let value = cd.eval_witness(&witness_name, witness_offset).to_bytes_le();
+        let mut buffer = [0u8; 64];
+        buffer[..value.len()].copy_from_slice(&value);
+        let value = Fr::from_uniform_bytes(&buffer);
```

Comment:
What's the type of `value`? Is this the easiest way to do the type conversion? Don't we have a routine already somewhere here?

Reply:
I didn't find any :/ It converts from `Bn254Field` to `halo2_curves::bn256::Fr` by going to little endian bytes and back.

Reply:
But shouldnt we already have that somewhere here for the witness? I'll take a look in a sec

Reply:
Ah the one I had in mind is from constants and not FieldElement